feat(training): flexible timestepping for forecaster (#1070)
ecmwf/anemoi-core: training/docs/modules/tasks.rst

   ``multistep_input`` outputs are fed into the next rollout step.

******************
 OffsetForecaster
******************

:class:`~anemoi.training.tasks.forecasting.OffsetForecaster` is an

> experimental generalization of

Reviewer: ```suggestion
experimental generalisation of
```

:class:`~anemoi.training.tasks.forecasting.Forecaster` that declares the

> sample directly through time offsets, i.e. times relative to the start of the forecast:

Reviewer: ```suggestion
sample directly through time offsets, i.e. times relative to the forecast initialisation time:
```

- ``input_offsets`` — input time offsets as frequency strings, e.g.

>   ``["-6H", "0H"]``.

Reviewer: I would put an example here of a different frequency other than hours

> - ``output_offsets`` — target time offsets, e.g. ``["6H"]``.

Reviewer: ```suggestion
- ``output_offsets`` — target time offsets as frequency strings, e.g. ``["6H"]``.
```

Reviewer 2: Is anything "hardcoding" hours at the moment? Assuming a 10minute gridded dataset, would something not work?

Reviewer: you're right it doesn't need to be hours

Author: Indeed, there is even a unit test using minutes.

- ``rollout`` / ``validation_rollout`` — as for
  :class:`~anemoi.training.tasks.forecasting.Forecaster`.

> 

Reviewer: I would put a heading here like 
"Examples"

Basic forecaster
================

The standard forecaster (two input steps, one output step, ``6H``
timestep) written with offsets:

.. code:: yaml

  task:
    _target_: anemoi.training.tasks.OffsetForecaster
    input_offsets: ["-6H", "0H"]

>     output_offsets: ["6H"]

Reviewer: could you add rollout here too?

- ``input_offsets`` — input time offsets as frequency strings, e.g.
  ``["-1H", "-30m", "0m"]``.
- ``output_offsets`` — target time offsets as frequency strings, e.g. ``["30M",  "1h"]``.

- ``rollout`` / ``validation_rollout`` — as for
  :class:`~anemoi.training.tasks.forecasting.Forecaster`.

Examples
========

Basic forecaster
----------------

The standard forecaster (two input steps, one output step, ``6H``
timestep) written with offsets:

.. code:: yaml

  task:
    _target_: anemoi.training.tasks.OffsetForecaster
    input_offsets: ["-6H", "0H"]
    output_offsets: ["6H"]
    rollout:
      start: 1
      epoch_increment: 0
      maximum: 1

Mixed input/output frequencies
------------------------------

Read two 6-hourly steps and predict the next six hours at 1-hourly
resolution:

.. code:: yaml

  task:
    _target_: anemoi.training.tasks.OffsetForecaster
    input_offsets: ["-6H", "0H"]
    output_offsets: ["1H", "2H", "3H", "4H", "5H", "6H"]
    rollout:
      start: 1
      epoch_increment: 1
      maximum: 6

The rollout shift
=================

The model can forecast arbitrarily far into the future by rollout: it is

> applied repeatedly, each step feeding its own outputs (and inputs) back as the inputs of

Reviewer: ```suggestion
applied repeatedly, each step feeding its own outputs back as the inputs of
```

Author: Wouldn't we like to be precise? Typically part of the input of a forecast step is reused in the next step upon rollout. (e.g. -6 0 -> 6, 0 6 -> 12 etc)
